packages/platforms/react-native/lib/retry-queue/file-based.ts

```diff
@@ -0,0 +1,146 @@
+function getLatestSpan (body: DeliveryPayload): DeliverySpan | undefined {
+  let latestSpan: DeliverySpan | undefined
+  let biggestTimestamp: bigint | undefined
+
+  for (const resourceSpan of body.resourceSpans) {
+    for (const scopeSpan of resourceSpan.scopeSpans) {
+      for (const span of scopeSpan.spans) {
+        if (biggestTimestamp === undefined || BigInt(span.endTimeUnixNano) > biggestTimestamp) {
+          latestSpan = span
+
+          // store the biggest timestamp separately as a bigint so we don't need
+          // to parse it on every iteration
+          biggestTimestamp = BigInt(latestSpan.endTimeUnixNano)
+        }
+      }
+    }
+  }
+
+  return latestSpan
+}
+
+function isValidFilename (filename: string): boolean {
+  // if the filename is too short then it can't be valid
+  if (filename.length < MININUM_FILENAME_LENGTH) {
+    return false
+  }
+
+  const timestamp = timestampFromFilename(filename)
+
+  // if we can't parse a timestamp then the filename can't be valid
+  if (!timestamp) {
+    return false
+  }
+
+  const nowInNanoseconds = BigInt(Date.now()) * NANOSECONDS_IN_MILLISECONDS
+
+  // files that are older than 24 hours are not valid as the data may not be
+  // relevant anymore
+  if (timestamp < nowInNanoseconds - NANOSECONDS_IN_DAY) {
+    return false
+  }
+
+  // files that have a timestamp more than 24 hours in the future are not
+  // valid as something may have gone wrong with the clock and this is
+  // unlikely to be a timezone issue
+  if (timestamp > nowInNanoseconds + NANOSECONDS_IN_DAY) {
+    return false
+  }
+
+  return true
+}
+
+// the minimum possible length of a payload filename (i.e. a 1 character
+// timestamp and span ID)
+const MININUM_FILENAME_LENGTH = 'retry-0-0.json'.length
+const NANOSECONDS_IN_MILLISECONDS = BigInt(1_000_000)
+const NANOSECONDS_IN_DAY = BigInt(24 * 60 * 60_000) * NANOSECONDS_IN_MILLISECONDS
+
+// the outcome of flushing a single file — either delete it (e.g. success or
+// permanent failure) or leave it alone for the next flush (retryable failure)
+const enum FlushOutcome { DeleteFile, LeaveFile }
+
+export default class FileBasedRetryQueue implements RetryQueue {
+  private readonly delivery: Delivery
+  private readonly directory: Directory
+
+  constructor (delivery: Delivery, directory: Directory) {
+    this.delivery = delivery
+    this.directory = directory
+  }
+
+  add (payload: TracePayload, _time: number): void {
+    const span = getLatestSpan(payload.body)
+
+    if (!span) {
+      return
+    }
+
+    // we use the latest span's timestamp in the filename so we can decide
+    // whether to keep or discard the span without parsing the payload
+    // as we can't be sure of nanosecond precision in JS environments, we also
+    // append the span ID so file names are unique across batches
+    const filename = `retry-${span.endTimeUnixNano}-${span.spanId}.json`
+
+    this.directory.write(filename, JSON.stringify(payload))
+      .catch(() => {})
```

Comment:
should we at least log a warning if this fails?

Reply:
@gareththackeray ?

Reply:
I would have thought so. Are there reasons why we might not want to?

Reply:
Only that you probably couldn't do anything about it and by default they'd just be logged to the end user's device so wouldn't be visible to the developer